metadoc-cli/src/main/scala/metadoc/cli/TermDisplay.scala

```diff
@@ -0,0 +1,510 @@
+
+            doneQueue.clear()
+
+            val dw = downloads.toVector
+              .map { url =>
+                url -> infos.get(url)
+              }
+              .sortBy { case (_, info) => -info.fraction.sum }
+
+            (q, dw)
+          }
+
+          for ((url, info) <- done0 ++ downloads0) {
+            assert(info != null, s"Incoherent state ($url)")
+
+            truncatedPrintln(url)
+            out.clearLine(2)
+            out.write(s"  ${info.display()}\n")
+          }
+
+          val displayedCount = (done0 ++ downloads0).length
+
+          if (displayedCount < lineCount) {
+            for (_ <- 1 to 2; _ <- displayedCount until lineCount) {
+              out.clearLine(2)
+              out.down(1)
+            }
+
+            for (_ <- displayedCount until lineCount)
+              out.up(2)
+          }
+
+          for (_ <- downloads0.indices)
+            out.up(2)
+
+          out.left(10000)
+
+          out.flush()
+          Thread.sleep(refreshInterval)
+          updateDisplayLoop(downloads0.length)
+      }
+    }
+
+    @tailrec private def fallbackDisplayLoop(previous: Set[String]): Unit =
+      Option(q.poll(100L, TimeUnit.MILLISECONDS)) match {
+        case None => fallbackDisplayLoop(previous)
+        case Some(Message.Stop) => // poison pill
+
+          // clean up display
+          for (_ <- 1 to 2; _ <- 0 until currentHeight) {
+            out.clearLine(2)
+            out.down(1)
+          }
+          for (_ <- 0 until currentHeight) {
+            out.up(2)
+          }
+
+        case Some(Message.Update) =>
+          val downloads0 = downloads.synchronized {
+            downloads.toVector
+              .map { url =>
+                url -> infos.get(url)
+              }
+              .sortBy { case (_, info) => -info.fraction.sum }
+          }
+
+          var displayedSomething = false
+          for ((url, info) <- downloads0 if previous(url)) {
+            assert(info != null, s"Incoherent state ($url)")
+
+            val (url0, extra0) = reflowed(url, info)
+
+            displayedSomething = true
+            out.write(s"$url0 $extra0\n")
+          }
+
+          if (displayedSomething)
+            out.write("\n")
+
+          out.flush()
+          Thread.sleep(fallbackRefreshInterval)
+          fallbackDisplayLoop(previous ++ downloads0.map {
+            case (url, _) => url
+          })
+      }
+
+    override def run(): Unit = {
+
+      Terminal.consoleDim("cols") match {
+        case Some(cols) =>
+          width = cols
+          out.clearLine(2)
+        case None =>
+          fallbackMode = true
+      }
+
+      if (fallbackMode)
+        fallbackDisplayLoop(Set.empty)
+      else
+        updateDisplayLoop(0)
+    }
+  }
+
+}
+
+object Cache {
+  trait Logger {
+    def foundLocally(url: String, f: File): Unit = {}
+    def startTask(url: String, file: File): Unit = {}
+    def taskProgress(url: String, downloaded: Long): Unit = {}
```

Comment:
It looks like only `startTask` and `taskProgress` is used. Can the rest be removed or we want to keep it?

Reply:
Removed unused methods, added a call to `completedTask`.